### DISCUSSION TITLE
Typical way to do green screen (Chroma Key) removal and composite onto a background

### DISCUSSION THREAD

**Discussion by GenericJam** [libvips/libvips#3097] [General]
I'm trying to  use Vix in Elixir to do green screen removal. Is there a normal way to do this? Everything I've tried so far has hit some sort of road block. I suspect this is a solved problem just not published. If someone could write how you'd do this in some other language like Python I can try to transcribe it.

Finished:
![image (3)](https://user-images.githubusercontent.com/5807145/196058610-79bc8daf-1ac3-4b90-a850-385b0817c746.png)
Sources:
![greenscreen](https://user-images.githubusercontent.com/5807145/196058643-66bcb087-42fb-480d-8b7e-2153eb180f2a.jpg)
![background](https://user-images.githubusercontent.com/5807145/196058661-304d8745-bbeb-4aa6-a7d1-b9fc559c1225.jpg)

**Comment by GenericJam** [comment]
To see the code and approach I'm already taking see here. https://stackoverflow.com/questions/74090698/chroa-key-removal-using-elixir-vix-libvips

**Reply by jcupitt** [reply]
Hi @GenericJam,

I'd do something like:

```python
#!/usr/bin/python3

import sys
import pyvips

foreground = pyvips.Image.new_from_file(sys.argv[1])
background = pyvips.Image.new_from_file(sys.argv[2], access="sequential")

# average a 10x10 square at the top left to guess the key colour
# this will be an RGB triple eg. [10, 10, 240]
key_colour = [i.avg() for i in foreground.crop(0, 0, 10, 10).bandsplit()]

print(f"key_colour = {key_colour}")

# find distance from key colour, then threshold
threshold = 20
mask = ((foreground - key_colour) ** 2).bandmean() > 3 * threshold ** 2

# set as the alpha and composite
result = background.composite(foreground.bandjoin(mask), "over")

result.write_to_file(sys.argv[3])
```

ie. find the RGB triple for the key, then compute a distance metric and threshold that.

You could improve it a bit by eroding the mask to remove the fringing, and maybe using CIELAB.

**Comment by GenericJam** [comment]
Thanks so much for your response, My previous effort was failing as I didn't realize opening the files differently made a difference in processing them later on.

**Reply by kipcole9** [reply]
The [Image](https://github.com/kipcole9/image) library for [Elixir](https://elixir-lang.org) now implements a [chroma_key](https://github.com/kipcole9/image/blob/main/lib/image.ex#L1198-L1261) function that implements both @jcupitt's threshold approach above, and a color range approach. Thanks to @GenericJam for the inspiration and collaboration and @jcupitt for the amazing `libvips`.

**Reply by jcupitt** [reply]
Yes, the `sequential` hint lets the libvips IO system open the image in streaming mode. This saves memory and is faster, but you can only make a single pass through the image (obviously), so no operations like 90 degree rotate.

There's a chapter in the docs about it with some background:

https://www.libvips.org/API/current/How-it-opens-files.html

**Reply by kipcole9** [reply]
In `Image` I default to `sequential` access (not `random`) which is why @GenericJam originally had issues - and I need to improve my documentation and error messaging.

**Reply by jcupitt** [reply]
I think I'd default to `random`. 

You can only use `seq` in some cases, so (IMO) it's best as an optimisation you can enable rather than default behaviour.
